text/0000-write-pointers.md

```diff
@@ -0,0 +1,463 @@
+    // Error, bar is not completely initialized (Bar.e is not initialized)
+}
+
+fn invalid_init_bar_v2(bar: &uninit Bar) {
+    bar.d = 10;
+    if bar.d == 9 {
+        return; // Error, bar is not completely initialized (Bar.e is not initialized)
+    }
+    bar.e = 10.0;
+}
+```
+
+If a closure captures a `&uninit T`, then it becomes a `FnOnce`, because of the write semantics, the destructors will not be run the first time.
+
+```Rust
+let x: Foo;
+
+let init = || x.a = 12; // init: FnOnce()  -> ()
+```
+
+**Note on Panicky Functions:**
+If a function panics, then all fields initialized in that function will be dropped. No cross-function analysis will be done.
+
+## `&out T`
+
+Using `&out T`, we can directly initialize a value and guarantee to write only behavior. \
+That would add give a memory location to write to directly instead of relying on move-elimination optimizations.
+
+```Rust
+#[derive(Clone, Copy)]
+struct Rgb(pub u8, pub u8, pub u8);
+/// This abstraction that exposes a Frame Buffer allocated by the OS, and is unsafe to read from
+struct FrameBuffer( ... );
+
+impl FrameBuffer {
+    /// initializes the FrameBuffer in place
+    fn new(&uninit self) { ... }
+
+    /// gets a write only refernce to pixel at position (row, col)
+    fn write_to_pixel(&mut self, row: usize, col: usize) -> &out Rgb {
+         ...
+    }
+}
+```
+This could be used like this
+```Rust
+let buffer;
+FrameBuffer::new(&uninit buffer);
+
+*buffer.write_to_pixel(0, 0) = Rgb(50, 50, 255);
+*buffer.write_to_pixel(10, 20) = Rgb(0, 250, 25);
+/// ...
+```
+
+**Note:** `Rgb` is `Copy`, if it wasn't we could not gaurentee that we can safely overwrite it
+
+## Constructors and Direct Initialization
+
+Using `&uninit` we can create constructors for Rust!
+```Rust
+struct Rgb(u8, u8 ,u8);
+
+impl Rgb {
+    fn init(&uninit self, r: u8, g: u8, b: u8) {
+        self.0 = r;
+        self.1 = g;
+        self.2 = b;
+    }
+}
+
+let color: Rgb;
+color.init(20, 23, 255);
+```
+
+and we can do direct initialization
+```Rust
+impl<T> Vec<T> {
+    pub fn emplace_back(&mut self) -> &uninit T {
+        ... // magic to allocate space and create reference
+    }
+}
+```
+
+and maintain write-only buffers
+```Rust
+struct WriteOnly([u8; 1024]);
+
+impl WriteOnly {
+    pub fn write(&out self, byte: u8, location: usize) {
+        self.0[location] = byte; // currently not possible to index like this, but we could imagine a IndexOut, that will handle this case
+    }
+}
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+**NOTE** This RFC does NOT aim to create new raw pointer types, so no `*out T` or `*uninit T`. There is no point in creating these.
+
+## Rules of `&uninit T`
+
+`&uninit T` should follow some rules in so that is is easy to reason about `&uninit T` locally and maintain soundness
+- `&uninit T` follows the same rules as `&mut T` for the borrow checker
+- `&uninit T` can only be assigned to once
```

Comment:
Ok, I didn't know that move checking is part of the borrow checker.

But the idea is that `&uninit` should an exclusive pointer, like `&mut`. Also after the first assign, `&uninit T` gets coerced into `&T` and `&mut T`, so it can be used as such.